src/main/java/io/divolte/server/recordmapping/Digester.java

```diff
@@ -0,0 +1,193 @@
+ */
+
+package io.divolte.server.recordmapping;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.reflect.TypeToken;
+import io.divolte.server.DivolteEvent;
+import io.divolte.server.recordmapping.DslRecordMapping.ValueProducer;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+@ParametersAreNonnullByDefault
+public final class Digester<T> {
+    private final String digestIdentifier;
+    private final Supplier<T> digestFactory;
+    private final Function<T,Consumer<ByteBuffer>> digestUpdaterFactory;
+    private final Function<T,Supplier<byte[]>> digestFinalizerFactory;
+    private List<ValueProducer<ByteBuffer>> digestPieces = new ArrayList<>();
+
+    private Digester(final String identifier,
+                     final Supplier<T> digestFactory,
+                     final Function<T,Consumer<ByteBuffer>> digestUpdater,
+                     final Function<T,Supplier<byte[]>> digestFinalizer) {
+        digestIdentifier = Objects.requireNonNull(identifier);
+        this.digestFactory = Objects.requireNonNull(digestFactory);
+        this.digestUpdaterFactory = Objects.requireNonNull(digestUpdater);
+        this.digestFinalizerFactory = Objects.requireNonNull(digestFinalizer);
+    }
+
+    public ValueProducer<ByteBuffer> result() {
+        final String identifier =
+            digestPieces.stream()
+                        .map(s -> String.format(".add(%s)", s))
+                        .collect(Collectors.joining("", digestIdentifier, ".result()"));
+        return new BytesValueProducer(identifier, this::calculateDigest);
+    }
+
+    private Optional<ByteBuffer> calculateDigest(final DivolteEvent e, final Map<String,Optional<?>> context) {
+        final T messageDigest = digestFactory.get();
+        final Consumer<ByteBuffer> digestUpdater = digestUpdaterFactory.apply(messageDigest);
+        final Supplier<byte[]> digestFinalizer = digestFinalizerFactory.apply(messageDigest);
+        digestPieces.stream()
+                    .map(x -> x.produce(e, context))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .forEachOrdered(digestUpdater);
+        return Optional.of(ByteBuffer.wrap(digestFinalizer.get()));
+    }
+
+    public Digester add(final ValueProducer<?> piece) {
+        final TypeToken<?> producerType = piece.producerType;
```

Comment:
Ugh, can't we pass a typehint to the function or something.

Reply:
The hint might lie. No lovely implicits to enforce it, and type erasure… it all just sucks.